chainerx_cc/chainerx/cuda/memory_pool.h

```diff
@@ -53,13 +57,65 @@ class PinnedMemoryAllocator : public Allocator {
     void Free(void* ptr) noexcept override { cudaFreeHost(ptr); }
 };
 
+namespace cuda_internal {
+
+// A chunk that points to a device memory.
+//
+// A chunk might be a splitted memory block from a larger allocation.
+// The prev/next pointers construct a doubly-linked list of contiguous memories
+// sorted by those base addresses.
+class Chunk {
+public:
+    Chunk(void* ptr, size_t offset, size_t bytesize)
+        : ptr_{reinterpret_cast<void*>(reinterpret_cast<intptr_t>(ptr) + offset)},  // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+          bytesize_{bytesize} {}
+    Chunk(const Chunk&) = default;
+    ~Chunk() {}
+
+    // Splits this chunk into a chunk with the given bytesize and one with the remaining.
+    //
+    // Modifies the bytesize and the next pointer of this chunk, and returns the chunk for the remaining.
+    std::unique_ptr<Chunk> Split(size_t bytesize);
+
+    // Merges this chunk with the next.
+    void MergeWithNext();
+
+    void SetPrev(Chunk* prev) { prev_ = prev; }
+    void SetNext(Chunk* next) { next_ = next; }
+
+    void* ptr() const { return ptr_; }
+    size_t bytesize() const { return bytesize_; }
+    const Chunk* prev() const { return prev_; }
+    Chunk* prev() { return prev_; }
+    const Chunk* next() const { return next_; }
+    Chunk* next() { return next_; }
+
+private:
+    void* ptr_{nullptr};  // Memory address.
+    size_t bytesize_{0};  // Chunk bytesize.
+    Chunk* prev_{nullptr};  // Prev memory pointer if splitted from a larger allocation
+    Chunk* next_{nullptr};  // Next memory pointer if splitted from a larger allocation
+};
+
+using FreeList = std::vector<std::unique_ptr<Chunk>>;  // List of free chunks with the same size.
+
+}  // namespace cuda_internal
+
+namespace {
```

Comment:
Unnamed scope is not available in header.

Reply:
https://ttsuki.github.io/styleguide/cppguide.ja.html#Unnamed_Namespaces_and_Static_Variables